src/librustc/middle/resolve_lifetime.rs

```diff
@@ -83,32 +82,37 @@ pub enum Region {
     Free(DefId, /* lifetime decl */ DefId),
 }
 
+fn new_region(hir_map: &Map, param: &GenericParam) -> (DefId, LifetimeDefOrigin) {
+    let def_id = hir_map.local_def_id(param.id);
+    let origin = match param.kind {
+        GenericParamKind::Lifetime { in_band, .. } => {
+            LifetimeDefOrigin::from_is_in_band(in_band)
+        }
+        _ => bug!("expected a lifetime param"),
+    };
+    (def_id, origin)
+}
+
 impl Region {
-    fn early(
-        hir_map: &Map,
-        index: &mut u32,
-        def: &hir::LifetimeDef,
-    ) -> (hir::LifetimeName, Region) {
+    fn early(hir_map: &Map, index: &mut u32, param: &GenericParam) -> (ParamName, Region) {
         let i = *index;
         *index += 1;
-        let def_id = hir_map.local_def_id(def.lifetime.id);
-        let origin = LifetimeDefOrigin::from_is_in_band(def.in_band);
+        let (def_id, origin) = new_region(hir_map, param);
```

Comment:
Couldn't this be replaced by having `LifetimeDefOrigin::from_param` (that takes the `&GenericParam`) instead of `LifetimeDefOrigin::from_is_in_band`?